\documentclass[webpdf,contemporary,large,namedate]{oup-authoring-template}%

\usepackage{graphicx}
\usepackage{hyperref}
\usepackage{url}
\usepackage{tabularx}
\usepackage{amsmath}
%\usepackage[ruled,vlined]{algorithm2e}
%\newcommand\mycommfont[1]{\footnotesize\rmfamily{\it #1}}
%\SetCommentSty{mycommfont}
%\SetKwComment{Comment}{$\triangleright$\ }{}
\renewcommand{\ttdefault}{cmtt}

\DeclareMathOperator*{\argmax}{argmax}

\begin{document}
\journaltitle{Journal Title Here}
\DOI{DOI HERE}
\copyrightyear{2024}
\pubyear{2024}
\access{Advance Access Publication Date: Day Month Year}
\appnotes{Paper}
\firstpage{1}

\title[Pangene graphs]{Exploring gene content with pangenome gene graphs}
\author[1,2,$\ast$]{Heng Li}
\address[1]{Dana-Farber Cancer Institute, 450 Brookline Ave, Boston, MA 02215, USA}
\address[2]{Harvard Medical School, 10 Shattuck St, Boston, MA 02215, USA}
\corresp[$\ast$]{Corresponding author. \href{email:hli@ds.dfci.harvard.edu}{hli@ds.dfci.harvard.edu}}

\abstract{
\textbf{Motivation:}
The gene content of an organism regulates the biology of the organism
and varies between species and between individuals of the same species.
Although several tools have been developed to identify gene content changes in bacterial genomes,
no existing tools are applicable to collections of large Eukaryotic genomes such as the human pangenome.\vspace{0.5em}\\
\textbf{Results:}
We developed pangene, a computational tool to identify gene orientation, gene order
and gene copy-number changes in a collection of genomes.
Pangene aligns a set of input protein sequences to the genomes,
resolves redundancies between protein sequences
and constructs a gene graph with each genome represented as a walk in the graph.
It additionally characterizes subgraphs that encodes gene content chnages.
Applied to the human pangenome, pangene identifies known gene-level variations
and reveals complex haplotypes that are not well studied before.
Pangene also works with long-read bacterial pangenome and reports similar numbers of core and accessory genes
in comparison to existing tools.\vspace{0.5em}\\
\textbf{Availability and implementation:}
\href{https://github.com/lh3/pangene}{https://github.com/lh3/pangene}
}

\maketitle

\section{Introduction}

A human genome contains about 20,000 protein-coding genes.
A small fraction of them have frequent copy-number or gene order changes in the human population~\citep{Sudmant:2010aa,Handsaker:2015ur}.
These genes are often under fast evolution
and may be responsible for immune responses,
affecting brain functionality~\citep{Ju:2016aa} and drug metabolism~\citep{Taylor:2020aa},
or associated with known diseases~\citep{Mercuri:2022aa}.
They may have profound biological and biomedical implications.

Thanks to the recent advances in sequencing technologies~\citep{Wenger_2019} and assembly algorithms~\citep{Nurk:2020we,Cheng:2021aa,Rautiainen:2023aa},
we can routinely derive haplotype-resolved assembly of genes under copy-number or order changes
and we have developed algorithms to construct a pangenome graph encoding the variations between genomes.
However, how to automatically identify these genes remains an unsolved problem.
Among the three pangenome construction tools used by the Human Pangenome Reference Consortium (HPRC),
minigraph~\citep{Li:2020aa} and minigraph-cactus~\citep{Hickey:2023aa} are
unable to align through complex genomic regions and may miss genes in long segmental duplications;
PGGB~\citep{Garrison2023.04.05.535718} collapses paralogous genes which makes it difficult to study individual paraplog.
In addition, all three tools do not reveal how genomic variations affect genes.
To study gene-level variations, HPRC
had to manually annotate genes on each haplotype~\citep{Liao:2023aa} which is a time-consuming process.
The current human pangenome tooling is not designed for studying gene variations.

In contrast, research in bacterial pangenome
focuses on protein-coding genes instead of genome sequences, to the point that
in the literature, a bacterial ``pangenome'' often refers to the collection of protein-coding genes.
Several high-quality tools have been developed for constructing the gene content of bacterial genomes~\citep{Page:2015aa,Ding:2018aa,Tonkin-Hill:2020aa,Gautreau:2020aa,Zhou:2020aa}.
In a nutshell, they start with \emph{ab initio} gene annotation in each genome,
cluster the restulted protein sequences,
and then post-process clusters to identify orthoglous genes
and to fix issues caused by imperfect assembly, annotation or clustering~\citep{Tonkin-Hill:2023aa}.
These bacterial pangenome tools however have not considered splicing,
multiple isoforms, frequent segmental duplications and the much larger size of the human genome.
They have not been shown to work with human pangenome data.

We therefore developed pangene, a new computational tool to explore the gene content of a pangenome.
Unlike bacterial pangenome pipelines, pangene effectively annotates protein-coding genes
by aligning protein sequences to each genome with miniprot~\citep{Li:2023ac}.
As miniprot can align through in-frame stop codons and frameshifts,
this procedure simplifies gene annotation and is robust to insertion/deletion errors in the input genomes.
Furthermore, pangene constructs a bidirected gene graph and can capture inversions
missed by bacterial pangenome tools.
It also provides an approximate algorithm to directly
identify gene copy-number or gene order variations.
Pangene is optimized for human genomes and also works for bacterial pangenome construction.

\section{Methods}

Pangene takes a set of protein sequences and multiple genome assemblies as input,
and outputs a graph in the GFA format~\citep{Li:2020aa}.
It involves two steps: aligning the set of protein sequences to each input assembly with miniprot~\citep{Li:2023ac},
and deriving a graph from the alignment with each contig encoded as a walk of genes.
Pangene provides utilities to classify genes into core genes that are present in most of the input genomes, or accessory genes otherwise.
Pangene can also identify generalized ``bubbles'' in the graph, which represent gene order,
gene copy-number or gene orientation variations among the input genomes.

Given perfect gene annotations, the pangene graph construction algorithm is conceptually simple:
it takes each gene as a node and adds an edge between two genes if they are adjacent on a genome (Fig.~\ref{fig:ex1}).
The practical difficulty is to obtain accurate annotation given redundant genes, paralogous genes and errors in assembly or alignment.
On the theoretical side, bubble finding is an unsolved problem on its own
and will also be a major topic of this article.

\begin{figure}[b!]
\centering
\includegraphics[width=.48\textwidth]{fig1}
\caption{Examples of pangene graphs. {\bf (a)} Human haplotypes around the
\emph{HLA-DRB1} gene. {\bf (b)} The pangene graph around \emph{HLA-DRB1}. {\bf
(c)} Human haplotypes around the \emph{RHD} gene. \emph{RHD} has copy-number
changes and \emph{TMEM50A} may be inverted. {\bf (d)} The corresponding pangene
graph.}\label{fig:ex1}
\end{figure}

\subsection{Defining pangene graphs}

\begin{figure}[t!]
\centering
\includegraphics[width=.48\textwidth]{fig2}
\caption{An example of a gene graph. {\bf (a)} A gene graph for visualization.
{\bf (b)} The corresponding GFA format.
{\bf (c)} Bidirected graph.
{\bf (d)} Biedged graph.
{\bf (e)} Directed graph.
{\bf (f)} Net graph, which loses information.
{\bf (g)} A depth-first traversal of the net graph with ``0'' representing a super node
not in the original net graph.}\label{fig:exa1}
\end{figure}

Let $V$ be the set of genes
and $X=V\times\{{\rm >},{\rm <}\}$ be the set of oriented genes.
If $v\in V$ is a gene, $x=(v,{\rm >})\in X$, or simply $x={\rm >}v$, denotes an oriented gene.
$\overline{x}$ gives the reverse complement of $x$.
i.e., if $x=(v,{\rm >})$, $\overline{x}=(v,{\rm <})$, and vice versa.
Throughout this article, we will use symbol $u$, $v$ or $w$ to denote a gene
and use $x$, $y$ or $z$ to denote an oriented gene.

A pangene graph is usually visualized as Fig.~\ref{fig:exa1}a~\citep{Wick:2015qf}
and can be mathematically defined in a few equivalent ways.
It can be formulated as a directed graph $G_D=(X,E)$ where $E\subset X\times X$ is the set of edges.
In the context of pangene graphs, $(x,y)$ is an edge if $x$ is followed by $y$ in an input genome,
or, due to the strand symmetry of DNA, $\overline{y}$ is followed by $\overline{x}$.
In the graph theory, this property is called the \emph{skew symmetry}.
Fig.~\ref{fig:exa1}e shows an example of a directed pangene graph
which can be described in the GFA format (Fig.~\ref{fig:exa1}b).

A pangene graph can also be thought as a bidirected graph $G_B=(V,E)$ where the definition of $E$ is the same as above.
An edge in $E$ is effectively associated with two directions.
For example, $({\rm >}v,{\rm <}w)\in E$ can be written as $v{\rm ><}w$,
suggesting the edge between $v$ and $w$ has a first direction ``${\rm >}$'' towards vertex $v$ and a second direction ``${\rm <}$'' towards vertex $w$.
Fig.~\ref{fig:exa1}c shows the bidirected equivalence of Fig.~\ref{fig:exa1}a.

In a directed graph, $x\to y\gets z$ is not a path because $x$ cannot reach $z$.
The similar constraint is applied to the bidirected graph $G_B$.
For example, $v{\rm >>}u{\rm <<}w$ is not a path because both $v$ and $w$ go into $u$.
$v{\rm >>}u{\rm ><}w$ is a path, which can also be expressed as a path in $G_D$: $({\rm >}v,{\rm >}u,{\rm <}w)$.

Biedged graph~\citep{Paten:2018aa} is a third way to formulate a pangene graph (Fig.~\ref{fig:exa1}d).
We will come back to this formulation after we describe the graph construction algorithm.

\subsection{Selecting proteins for alignment}

\subsection{Constructing a pangene graph}

\section{Results}

\subsection{Finding structural variants between two human genomes}

We downloaded GenCode comprehensive human annotation v44,
retained the protein coding genes and filtered out readthrough transcripts and mitochondrial genes.
Because GenCode does not include HLA-DRB3, HLA-DRB4 and several KIR genes,
we manually added 20 {\it HLA-DRB} genes from the IPD-IMGT/HLA database and
15 KIR genes from the IPD-KIR database.
In the end, we constructed a protein set with 109,317 proteins, representing 19,335 protein coding genes.

We aligned the proteins to the human reference genome GRCh38~\citep{Schneider:2017aa} and T2T-CHM13~\citep{Nurk:2022up}
with miniprot~\citep{Li:2023ac} v0.12 under option ``{\tt --outs=0.97 -Iu}''
and constructed a pangene graph.
The resulting graph contains 19,058 genes and 19,361 edges.
We identified 104 large-scale variants, represented by ``bubbles'' in the graph,
involving 1,026 protein coding genes affected by gene copy, gene order or gene orientation changes between the two genome.
These genes include many known events such as {\it PDPR}, {\it SMN2}, {\it CTAGE9}, {\it HPR}, {\it ORM1}, {\it CCL4}, {\it NCF1} and Amylase~\citep{Handsaker:2015ur,Sudmant:2010aa}
as well as new events.
We manually checked the miniprot alignment of 20 relatively small events
and believed pangene is reporting the desired haplotype structures.
Some of the large gene clusters affected by structural changes, such as SMN2 and Amylase,
would not be easily captured by whole-genome alignment as they are not represented by colinear alignments.

\begin{table}[!tb]
\caption{Bacterial pangenome analysis\label{tab:bac}}
%\begin{tabular}{lrrr}
\begin{tabular*}{\columnwidth}{@{\extracolsep\fill}lrrr@{\extracolsep\fill}}
\toprule
%& Panaroo & Panaroo & pangene & PPanGGOLiN\\
%&         & (merge) \\
%\midrule
%Mtb: \#total genes   & 4,261  & 4,216  & 4,217  & 4,742  \\
%Mtb: \#core genes    & 3,664  & 3,657  & 3,646  & 3,465  \\
%Ecoli: \#total genes & 14,152 & 13,723 & 13,009 & 14,630 \\
%Ecoli: \#core genes  & 2,890  & 2,893  & 3,081  & 2,966  \\
& Panaroo & pangene & PPanGGOLiN\\
\midrule
Mtb: \#total genes   & 4,216  & 4,217  & 4,742  \\
Mtb: \#core genes    & 3,657  & 3,646  & 3,465  \\
Ecoli: \#total genes & 13,723 & 13,009 & 14,630 \\
Ecoli: \#core genes  & 2,893  & 3,081  & 2,966  \\
\botrule
\end{tabular*}
\begin{tablenotes}
\item Panaroo, pangene and PPanGGOLiN were applied to two sets of gapless bacterial assemblies:
146 \emph{M. tuberculosis} (Mtb) strains and 50 \emph{E. coli} strains.
Mtb genes were annotated by Prokka and Ecoli genes were annotated by NCBI and downloaded from GenBank.
A ``core'' gene is a gene that is inferred to be present in all assemblies in each dataset.
Panaroo was invoked in the strict mode with paralogs merged (option ``{\tt
--clean-mode strict --merge\_paralogs}'').
\end{tablenotes}
\end{table}

\subsection{Analyzing 98 human haplotypes}

\subsection{Analyzing 146 \textit{M. tuberculosis} strains}

We downloaded the \emph{M. tuberculosis} reference strain H37Rv and its gene annotation from RefSeq (AC:GCF\_000195955.2),
which included 3,906 protein sequences.
We obtained the complete long-read assemblies of 145 other strains from \citet{Marin:2022aa}.
Following the instruction of Panaroo~\citep{Tonkin-Hill:2020aa},
we ran Prodigal~\citep{Hyatt:2010aa} v2.6.3 on the reference strain to train the Prodigal model
and ran Prokka~\citep{Seemann:2014aa} v1.14.6 with the pretrained model to predict protein coding genes
in the 145 non-reference strains.
We used CD-HIT~\citep{Li:2006aa,Fu:2012aa} v4.8.1 with option ``{\tt -c 0.98}'' to cluster non-reference protein sequences,
which resulted in 6,750 clusters.
We mapped these proteins to each \emph{M. tuberculosis} genome using miniprot with option ``{\tt -S}'' to disable splicing.
We finally ran pangene with ``{\tt -p.001}''
to keep all genes regardless of their frequency in the pangenome.

Pangene constructed a graph consisting of 4,217 genes,
3,646 of which were considered to be present in all 146 genomes (Table~\ref{tab:bac}).
To check if pangene captured the gene content in these strains,
we compared the pangene result to Panaroo v1.3.4 in the strict mode.
We aligned the Panaroo proteins to the pangene proteins with mmseqs2~\citep{Steinegger:2017aa} v13.45111
and identified 63 Panaroo proteins do not hit to pangene proteins.
We mapped the 63 proteins to H37Rv with miniprot
and found 60 of them can be aligned
and 76\% of the aligned regions overlap with annotated CDS in RefSeq.
Manually investigating the overlaps revealed that most of the 60 proteins
were aligned to the opposite strand of some RefSeq genes or in different reading frames.
Identifying homology based on the genomic locations of input proteins,
pangene did not include them into the final graph.
According to the RefSeq annotation,
only 0.2\% of coding regions in H37Rv are present in more than one genes --
overlapping genes are rare in \emph{M. tuberculosis}.

We additionally ran PPanGGOLiN~\citep{Gautreau:2020aa} v1.2.105 with the Prokka annotation as the input.
PPanGGOLiN collected 4,742 genes in the pangenome with 3,465 present in all.
279 genes did not hit to genes selected by pangene.
276 of these genes could be aligned to H37Rv by miniprot
and 90\% of the aligned regions overlap with annotated CDS in RefSeq.
PPanGGOLiN tends to select more overlapping genes in different reading frames,
comfirming the observation by~\citet{Tonkin-Hill:2020aa}.

\subsection{Analyzing 50 \textit{E. coli} strains}

\emph{M. tuberculosis} has low diversity with each strain similar to each other~\citep{Marin:2022aa}.
To understand how pangene performs given more diverse strains,
we downloaded the genomes 50 \emph{E. coli} strains with complete assemblies~\citep{Shaw:2021aa}.
We did not run Prokka on this dataset but instead used the gene annotation provided by NCBI.

We followed the same procedure to run pangenome tools.
To get clean graph, pangene by default filters out genes that had $>$10 edges or connected $>$3 distant loci in the graph.
This filter only removed six genes in \emph{M. tuberculosis} dataset but
it filtered out 627 genes in \emph{E. coli}.
We added option ``{\tt -p.001 -g50 -r10}'' to retain more genes.
Although pangene collected fewer genes (Table~\ref{tab:bac}),
only 98 Panaroo genes do not hit to genes collected by pangene.
The differences between the tools might be determined by subtle thresholds on how to resolve homologies
and may not reflect the capability of each algorithm.

\section{Discussions}

\section{Acknowledgements}

{\bf Funding:} NHGRI R01HG010040 and Chan-Zuckerberg Initiative

%\bibliographystyle{abbrvnat}
\bibliographystyle{apalike}
\bibliography{pangene}

\end{document}
