\documentclass{bioinfo}
\copyrightyear{2024}
\pubyear{2024}

\usepackage{graphicx}
\usepackage{hyperref}
\usepackage{url}
\usepackage{tabularx}
\usepackage{amsmath}
\usepackage[ruled,vlined]{algorithm2e}
\newcommand\mycommfont[1]{\footnotesize\rmfamily{\it #1}}
\SetCommentSty{mycommfont}
\SetKwComment{Comment}{$\triangleright$\ }{}
\renewcommand{\ttdefault}{cmtt}

\usepackage{natbib}
\bibliographystyle{apalike}

\DeclareMathOperator*{\argmax}{argmax}

\begin{document}
\firstpage{1}

\title[Pangene graphs]{Investigating gene content with pangenome gene graphs}
\author[Li]{Heng Li$^{1,2}$}
\address{$^1$Dana-Farber Cancer Institute, 450 Brookline Ave, Boston, MA 02215, USA,
$^2$Harvard Medical School, 10 Shattuck St, Boston, MA 02215, USA}

\maketitle

\begin{abstract}

\section{Motivation:}
The gene content of an organism regulates its biology to a large extent. The
gene content varies between species and between individuals of the same species
which distinguishes species and individuals. Although several tools have been
developed to explore the gene content changes in bacterial genomes, they are not applicable to rely
on accurate automated gene annotations

\section{Availability and implementation:}
\href{https://github.com/lh3/pangene}{https://github.com/lh3/pangene}

\section{Contact:} hli@ds.dfci.harvard.edu
\end{abstract}

\section*{Introduction}

\begin{methods}
\section*{Methods}

Pangene graph construction takes a set of protein sequences and multiple
genome assemblies as input, and outputs a graph in the GFA
format~\citep{Li:2020aa}. It involves two steps: aligning the set of protein
sequences to each input assembly with miniprot~\citep{Li:2023ac}, and deriving
a graph from the alignment with each contig encoded as a walk of genes.
Pangene provides utilities to classify genes into core genes that are present
in most of the input genomes, or accessory genes otherwise. Pangene can also
identify generalized ``bubbles'' in the graph, which represent gene order, gene
copy-number or gene orientation variations among the input genomes.

Given perfect gene annotations, the graph construction algorithm is
conceptually simple: pangene takes each gene as a node and adds an edge between
two genes if they are adjacent on a genome (Fig.~\ref{fig:ex1}). The caveat is
to handle inaccurate annotation which could be caused by assembly errors,
protein alignment errors and redundancy in input protein sequences. In
addition, bubble finding is a challenging theoretical problem on its own.

\subsection*{Notations}

\subsection*{Selecting proteins for alignment}

\subsection*{Constructing a pangene graph}

\begin{figure}
\centering
\includegraphics[width=.48\textwidth]{fig1}
\caption{Examples of pangene graphs. {\bf (a)} Human haplotypes around the
\emph{HLA-DRB1} gene. {\bf (b)} The pangene graph around \emph{HLA-DRB1}. {\bf
(c)} Human haplotypes around the \emph{RHD} gene. \emph{RHD} has copy-number
changes and \emph{TMEM50A} may be inverted. {\bf (d)} The corresponding pangene
graph.}\label{fig:ex1}
\end{figure}

\begin{figure}
\centering
\includegraphics[width=.48\textwidth]{fig2}
\caption{An example of sequence graph.}
\end{figure}

\begin{figure}
\centering
\includegraphics[width=.48\textwidth]{fig3}
\caption{An example of sequence graph.}
\end{figure}

\begin{figure}
\centering
\includegraphics[width=.48\textwidth]{fig4}
\caption{An example of sequence graph.}
\end{figure}

\end{methods}

\section*{Results}

\subsection*{Finding structural variants between two human genomes}

We downloaded GenCode comprehensive human annotation v44,
retained the protein coding genes and filtered out readthrough transcripts and mitochondrial genes.
Because GenCode does not include HLA-DRB3, HLA-DRB4 and several KIR genes,
we manually added 20 {\it HLA-DRB} genes from the IPD-IMGT/HLA database and
15 KIR genes from the IPD-KIR database.
In the end, we constructed a protein set with 109,317 proteins, representing 19,335 protein coding genes.

We aligned the proteins to the human reference genome GRCh38~\citep{Schneider:2017aa} and T2T-CHM13~\citep{Nurk:2022up}
with miniprot~\citep{Li:2023ac} v0.12 under option ``{\tt --outs=0.97 -Iu}''
and constructed a pangene graph.
The resulting graph contains 19,058 genes and 19,361 edges.
We identified 104 large-scale variants, represented by ``bubbles'' in the graph,
involving 1,026 protein coding genes affected by gene copy, gene order or gene orientation changes between the two genome.
These genes include many known events such as {\it PDPR}, {\it SMN2}, {\it CTAGE9}, {\it HPR}, {\it ORM1}, {\it CCL4}, {\it NCF1} and Amylase~\citep{Handsaker:2015ur,Sudmant:2010aa}
as well as new events.
We manually checked the miniprot alignment of 20 relatively small events
and believed pangene is reporting the desired haplotype structures.
Some of the large gene clusters affected by structural changes, such as SMN2 and Amylase,
would not be easily captured by whole-genome alignment as they are not represented by colinear alignments.

\begin{table}[!tb]
\processtable{Bacterial pangenome analysis}
{\footnotesize\label{tab:bac}
\begin{tabular}{p{2.3cm}rrrr}
\toprule
& Panaroo & Panaroo & pangene & PPanGGOLiN\\
&         & (merge) \\
\midrule
Mtb: \#total genes   & 4,261  & 4,216  & 4,217  & 4,742  \\
Mtb: \#core genes    & 3,664  & 3,657  & 3,646  & 3,465  \\
Ecoli: \#total genes & 14,152 & 13,723 & 13,009 & 14,630 \\
Ecoli: \#core genes  & 2,890  & 2,893  & 3,081  & 2,966  \\
\botrule
\end{tabular}
}{Panaroo, pangene and PPanGGOLiN were applied to two sets of gapless bacterial assemblies:
146 \emph{M. tuberculosis} (Mtb) strains and 50 \emph{E. coli} strains.
Mtb genes were annotated by Prokka and Ecoli genes were annotated by NCBI and downloaded from GenBank.
A ``core'' gene is a gene that is inferred to be present in all assemblies in each dataset.
Panaroo was invoked in both the strict mode (option ``{\tt --clean-mode strict}'')
and in the strict paralog merging mode (option ``{\tt --clean-mode strict --merge\_paralogs}'').}
\end{table}

\subsection*{Analyzing 98 human haplotypes}

\subsection*{Analyzing 146 \emph{M. tuberculosis} strains}

We downloaded the \emph{M. tuberculosis} reference strain H37Rv and its gene annotation from RefSeq (AC:GCF\_000195955.2),
which included 3,906 protein sequences.
We obtained the complete long-read assemblies of 145 other strains from \citet{Marin:2022aa}.
Following the instruction of Panaroo~\citep{Tonkin-Hill:2020aa},
we ran Prodigal~\citep{Hyatt:2010aa} v2.6.3 on the reference strain to train the Prodigal model
and ran Prokka~\citep{Seemann:2014aa} v1.14.6 with the pretrained model to predict protein coding genes
in the 145 non-reference strains.
We used CD-HIT~\citep{Li:2006aa,Fu:2012aa} v4.8.1 with option ``{\tt -c 0.98}'' to cluster non-reference protein sequences,
which resulted in 6,750 clusters.
We mapped these proteins to each \emph{M. tuberculosis} genome using miniprot with option ``{\tt -S}'' to disable splicing.
We finally ran pangene with ``{\tt -p.001}''
to keep all genes regardless of their frequency in the pangenome.

Pangene constructed a graph consisting of 4,217 genes,
3,646 of which were considered to be present in all 146 genomes (Table~\ref{tab:bac}).
To check if pangene captured the gene content in these strains,
we compared the pangene result to Panaroo v1.3.4 in the strict mode.
We aligned the Panaroo proteins to the pangene proteins with mmseqs2~\citep{Steinegger:2017aa} v13.45111
and identified 63 Panaroo proteins do not hit to pangene proteins.
We mapped the 63 proteins to H37Rv with miniprot
and found 60 of them can be aligned
and 76\% of the aligned regions overlap with annotated CDS in RefSeq.
Manually investigating the overlaps revealed that most of the 60 proteins
were aligned to the opposite strand of some RefSeq genes or in different reading frames.
Identifying homology based on the genomic locations of input proteins,
pangene did not include them into the final graph.
According to the RefSeq annotation,
only 0.2\% of coding regions in H37Rv are present in more than one genes --
overlapping genes are rare in \emph{M. tuberculosis}.

We additionally ran PPanGGOLiN~\citep{Gautreau:2020aa} v1.2.105 with the Prokka annotation as the input.
PPanGGOLiN collected 4,742 genes in the pangenome with 3,465 present in all.
279 genes did not hit to genes selected by pangene.
276 of these genes could be aligned to H37Rv by miniprot
and 90\% of the aligned regions overlap with annotated CDS in RefSeq.
PPanGGOLiN tends to select more overlapping genes in different reading frames,
comfirming the observation by~\citet{Tonkin-Hill:2020aa}.

\subsection*{Analyzing 50 \emph{E. coli} strains}

\emph{M. tuberculosis} has low diversity with each strain similar to each other~\citep{Marin:2022aa}.
To understand how pangene performs given more diverse strains,
we downloaded the genomes 50 \emph{E. coli} strains with complete assemblies~\citep{Shaw:2021aa}.
We did not run Prokka on this dataset but instead used the gene annotation provided by NCBI.

We followed the same procedure to run pangenome tools.
To get clean graph, pangene by default filters out genes that had $>$10 edges or connected $>$3 distant loci in the graph.
This filter only removed six genes in \emph{M. tuberculosis} dataset but
it filtered out 627 genes in \emph{E. coli}.
We added option ``{\tt -p.001 -g50 -r10}'' to retain more genes.
Although pangene collected fewer genes (Table~\ref{tab:bac}),
only 98 Panaroo genes do not hit to genes collected by pangene.
The differences between the tools might be determined by subtle thresholds on how to resolve homologies
and may not reflect the capability of each algorithm.

\section*{Discussions}

\section*{Acknowledgements}

{\bf Funding:} NHGRI R01HG010040 and Chan-Zuckerberg Initiative

\bibliography{pangene}

\end{document}
